PEP 795: Deep Immutability in Python (#4468)
python/peps: peps/pep-0795.rst

PEP: 795
Title: Deep Immutability in Python
Author: Matthew Johnson <matjoh@microsoft.com>,
        Matthew Parkinson <mattpark@microsoft.com>,
        Sylvan Clebsch <sylvan.clebsch@microsoft.com>,
        Fridtjof Peer Stoldt <fridtjof.stoldt@it.uu.se>,
        Tobias Wrigstad <tobias.wrigstad@it.uu.se>
Sponsor: Michael Droettboom <mike@droettboom.com>
Discussions-To: https://discuss.python.org/t/96014
Status: Draft
Type: Standards Track
Created: 19-Jun-2025
Python-Version: 3.15
Post-History:

Abstract
========

This PEP proposes adding a mechanism for deep immutability to
Python. The mechanism requires some changes to the core language,
but user-facing functions are delivered in a module called
``immutable``. This module provides the following functions and types:

1. The function ``freeze(obj)`` -- turns ``obj`` deeply immutable

2. The function ``isfrozen(obj)`` -- returns ``True`` if ``obj`` is immutable

3. The type ``NotFreezable`` which is an empty type which cannot be made immutable and can be used as a super class to classes whose instances should not be possible to freeze

> 4. The type ``NotFreezableError`` which is raised on an attempt to mutate an immutable object

Reviewer: Isn't this for an attempt to freeze and unfreezable object?  Below, it says it raises a `TypeError` when trying to mutate an immutable object.

           and what is shared.

See also the discussion about extensions further down in this
document.

Deep immutability can be implemented efficiently. An alternative approach
would be to detect data-races using a read-barrier based approach, however,
this cannot be implemented as efficiently. We discuss this in the alternatives
section. As highlighted above, immutability also has value in single-threaded
application, i.e. where there is no fear of data races.

Optimisations and Caching Benefits
----------------------------------

Immutable objects provide opportunities for optimisation, such as
structural sharing, memoization, and just-in-time (JIT)
compilation techniques (specialising for immutable data, e.g.
fixed shape, fewer barriers, inlining, etc.). Freezing objects can
allow Python to implement more efficient caching mechanisms and
enable compiler optimisations that rely on immutability
assumptions. This PEP will permit such opportunities to go
beyond today's immutable objects (like ``int``, ``string``) and
*shallow* immutable objects (``tuple``, ``frozenset``).

Specification
=============

Note: our current prototype implementation was authored on-top of

> Python 3.12. To avoid blocking on rebasing on 3.14 to force

Reviewer 2: This would be for 3.15 at this point.

   ``freeze`` behaves as in alternative 1), in the second case,
   it behaves as in alternative 2). We could further track whether
   an object is strictly immutable or not in order to prevent
   non-strictly immutable objects to participate in operations
   which require strictness. This adds additional complexity to
   the implementation, and also for the user.

This PEP proposes following alternative 1, where freezing either
succeeds or fails partially.

New Obligations on C Extensions
-------------------------------

Due to the opt-in decision, there are no *obligations* for C
extensions that do not want to add support for immutability.

Because our implementation builds on information available to the CPython
cycle detector, types defined through C code will support immutability
"out of the box" as long as they use Python standard types to store
data and uses the built-in functions of these types to modify the data.

To make its instances freezable, a type that uses C extensions
that adds new functionality implemented in C must register
themselves using ``register_freezable(type)``. Example:

.. code-block:: Python

> 

Reviewer 2: Should probably add a C API function for this too.

``foo`` gets its own copy of ``x`` which will have value of the enclosing
``x`` at the time of freezing, in this case 0.

Thus, the assignment ``x = 1`` is still permitted as it will not affect
``foo``, and it may therefore not contribute to a data race. Furthermore,
the result of calling ``foo()`` will be 0 -- not 1!

This is implemented by having ``x`` in ``foo`` point to a fresh
cell and then freezing the cell (and similar for global capture).
Note that this also prevents ``x`` from being reassigned.

We believe that this design is a sweet-spot that is intuitive and
permissive. Note that we will treat freezing functions that
capture enclosing state in the same way regardless of whether the
enclosing state is another function or the top-level (i.e., the
enclosing scope is ``globals()``).

(A **rejected alternative** is to freeze ``x`` in the
enclosing scope. This is problematic when a captured variable is
in ``globals()`` and also rejects more programs.)

Now consider freezing the following function:

.. code-block:: python

  from immutable import freeze

  def example2():
      x = 0
      def foo(a = False):
          nonlocal x

>             if a:

Reviewer 2: ```suggestion
          if a:
```

notably not contigent on :pep:`703`.

This work has three different components which we intend to
package into two discrete PEPs (called A and B below):

1. Support for identifying and freeing cyclic immutable garbage
   using reference counting. (PEP A)

2. Support for sharing immutable data across subinterpreters using
   atomic reference counting of immutable objects to permit
   concurrent increments and decrements on shared object RC's. (PEP A)

3. Support for sharing mutable data across subinterpreters, with
   dynamic ownership protecting against data races. (PEP B)

Together these components deliver "Data-Race Free Python".
Note that "PEP A" has value even if "PEP B" would not materialise
for whatever reason.

Simplified Garbage Collection for Immutable Object Graphs
---------------------------------------------------------

In `previous work <https://dl.acm.org/doi/10.1145/3652024.3665507>`_,
we have identified that objects that make up cyclic immutable
garbage will always have the same lifetime. This means that a
single reference count could be used to track the lifetimes of
all the objects in such a strongly connected component (SCC).

> We plan to extend the freeze logic with a SCC analysis that

Reviewer 2: ```suggestion
We plan to extend the freeze logic with an SCC analysis that
```

that a reference of a mutable type never points to an immutable
object, and conversely. These points will be unpacked and made
more clear in the PEP for the ownership model.

Reference Implementation
========================

`Available here <https://github.com/mjp41/cpython/pull/51>`_.

There are some discrepancies between this PEP and the reference
implementation, including:

- The ``NotFreezable`` type is currently freezable (but inheriting
  from it stops instances of the inheriting class from being made immutable).

Rebasing on Python 3.14
=======================

We have found two areas that need to be addressed to integrate this work with "free-threaded Python": data-representation and data-races during freeze.

Data-representation for immutability
------------------------------------

With free-threaded Python the representation of the reference
count has been changed. We could either borrow a bit to represent
if an object is immutable, or alternatively, we could use the new
``ob_tid`` field to have a special value for immutable state. Using
``ob_tid`` would allow for standard mutable thread local objects to
remain the fast path, and is our preferred alternative.

> The extensions use use SCC calculations to detect cycles in

Reviewer 2: ```suggestion
The extensions to use SCC calculations to detect cycles in
```

PEP: 795
Title: Deep Immutability in Python
Author: Matthew Johnson <matjoh@microsoft.com>,
        Matthew Parkinson <mattpark@microsoft.com>,
        Sylvan Clebsch <sylvan.clebsch@microsoft.com>,
        Fridtjof Peer Stoldt <fridtjof.stoldt@it.uu.se>,
        Tobias Wrigstad <tobias.wrigstad@it.uu.se>

> Sponsor: Michael Droettboom <mdroettboom@microsoft.com>

Reviewer 3: cc @mdboom for confirmation

Reviewer: Confirmed.

-  Focuses more on programming model, not just implementation details
-  Aims to be clearer about what the costs are for different parts of
   the design, both in terms of complexity for programmers and
   implementers of this PEP
-  Aims to make the semantics of freezing clearer, and provide more
   examples of how freezing propagates
-  Adds a new design that provides control over freeze propagation
-  Adds escape hatches and clearly points out who implements them
-  Motivates the design by linking it clearer to sharing objects across
   sub-interpreters
-  Aims to be clearer about the semantics of immutable functions (and why
   they are unavoidable)
-  Discusses the role of types and future plans for types
-  Includes direct sharing of immutable objects across sub-interpreters,
   rather than making a separate PEP for this
-  Removes a lot of the rejected alternatives from the original PEP;
   this is motivated by this PEP already being very long, and because
   the inclusion of direct sharing across sub-interpreters motivate many
   of the design decisions

We are very grateful to the many comments and threads on DPO that have

> contributed to this PEP.

Reviewer 4: This can move to an Acknowledgements section before Copyright at the end:
```suggestion
```

As explained above, that a function is immutable means that it is
safe to share across sub-interpreters. This means that the
function may not capture a direct reference to mutable state. The
function may be passed mutable state, and may get access to the
current interpreter's mutable state through module imports, but it
may not make one sub-interpreter's mutable state accessible to
another sub-interpreter.

A function can be made immutable either implicitly (*e.g.,*
because the type containing it is frozen) or explicitly (*e.g.,*
through a ``@frozen`` decorator or an explicit ``freeze()`` call).
When a function is frozen, we *decouple* it from its defining
context in two ways. This logic is handled transparently in the
pre-freeze hook of the function type object:

1. We re-point its captured globals to point to a copy that contains
   only the values that the function captured. (So ``random`` and ``x``
   in our example above, but not ``y``.) The copy is shallow.
2. We freeze the globals copy which freezes the *values* which
   are shared between the globals dictionaries, but keeps the
   variables in the defining context's globals dictionary mutable.

>    Thus, ``x = [5, 6, 7]`` is permissable, and because that updates

Reviewer 4: ```suggestion
   Thus, ``x = [5, 6, 7]`` is permissible, and because that updates
```

- The module object itself must me marked as freezable by setting freezable
  to either ``Yes``, ``Explicit`` or ``Proxy``
- Add write-barriers (``Py_CHECKWRITE(obj)``) that check for mutability
  before they modify the state of any C-object. This is only needed for
  C-objects which can be made immutable.

Note that the freezability also has an effect on how module state can be
accesses. A freezability of ``Yes`` and ``Explicit`` may result in multiple
sub-interpreters accessing the (frozen) module object at once. The ``Proxy``
mode will ensure that each mutable module object is only accessed by one
interpreter at a time.

C extensions are allowed to store mutable state, this is how escape
hatches can be implemented. However, they have to ensure that all
accesses to this data is *thread-safe* and that it does not cause
a mutable object from one sub-interpreter to become accessible to
another sub-interpreter.

Changes to the C ABI
====================

``Py_CHECKWRITE(obj)``
   This macro is used to check whether the argument (which must be
   castable to ``Py_Object *``) has its immutable bit set. If the
   argument is immutable, this evaluates to ``false``. If the argument

>    is writeable, it evalutes to ``true``.

Reviewer 4: ```suggestion
   is writeable, it evaluates to ``true``.
```

  objects. This requires changes in ~150 places in core
  Python. In particular, dictionaries, lists, and sets
  are updated to perform ``Py_CHECKWRITE`` calls in the
  appropriate places.
- Add additional synchronisation for updating the code
  objects of function objects due to optimisations in the
  interpreter. (Such synchronisation is already in place
  in free-threaded Python.)
- Make weak references freezable.

Performance of immutability
###########################

Below we share some *preliminary* numbers on performance based on
our current prototype. Note that we have not worked on optimisation
yet, our focus has been on correctness.

Effect on existing programs
===========================

To understand the impact of immutability on *existing* Python code
that does not use it, we compared the time it takes to run the
full CPython test suite with the time it takes to run it with our
changes on-top of it. Because of some intermittent tests failing
on the CPython baseline, these tests were performed with our
changes rebased on-top of CPython 3.14 (sha ``754e7c9b``). The
test suite contained 1109 Python files containing a total of

> 627917 LOC. While performance of test code is not

Reviewer 4: ```suggestion
test suite contained 1,109 Python files containing a total of
627,917 LOC. While performance of test code is not
```

necessarily indicative of performance of production code, the
breadth and volume of the code gives a good indication of how

> successfull we are in meeting our goal. On average over 10 runs on

Reviewer 4: ```suggestion
successful we are in meeting our goal. On average over 10 runs on
```

the ownership model that we envision can be found in our PLDI 2025:
`Dynamic Region Ownership for Concurrency
Safety <https://wrigstad.com/pldi2025.pdf>`__.

It is important to point out that data-race free Python is different
from free-threaded Python, but aims to be fully compatible with it, and
we believe that they can benefit from each other. In essence
free-threaded Python focuses on making the CPython run-time resilient
against data races in Python programs: a poorly synchronized Python
program should not be able to corrupt reference counts, or other parts
of the Python interpreter. The complementary goal pursued by DRF Python
is to make it impossible for Python programs to have data races. Support
for deeply immutable data can be seen as the first step towards this
goal.

Reference implementation
########################

Our `reference implementation`_, which we have made publicly available on
Github, implements all of the features described above.
We steal a bit from the top of the reference count in the object header. In 64-bit
python this comes from the (currently unused) top word and thus has no impact at

> all on the runtime. In 32-bit python, the second bit of the word is used (since negative

Reviewer 4: ```suggestion
Python this comes from the (currently unused) top word and thus has no impact at
all on the runtime. In 32-bit Python, the second bit of the word is used (since negative
```

PEP: 795
Title: Deep Immutability for Efficient Sharing and Concurrency Safety
Author: Matthew Johnson <matjoh@microsoft.com>, Matthew Parkinson <mattpark@microsoft.com>, Sylvan Clebsch <sylvan.clebsch@microsoft.com>, Fridtjof Stoldt <fridtjof.stoldt@it.uu.se>, Tobias Wrigstad <tobias.wrigstad@it.uu.se>
Sponsor: Michael Droettboom <mdboom@gmail.com>
Discussions-To: https://discuss.python.org/t/pep-795-revamped-deep-immutability-for-efficient-sharing-and-concurrency-safety/105356
Status: Draft
Type: Standards Track
Created: 19-Jun-2025

> Python-Version: 3.15

Reviewer 4: The 3.15 feature freeze has passed.
```suggestion
Python-Version: 3.16
```

Title: Deep Immutability for Efficient Sharing and Concurrency Safety
Author: Matthew Johnson <matjoh@microsoft.com>, Matthew Parkinson <mattpark@microsoft.com>, Sylvan Clebsch <sylvan.clebsch@microsoft.com>, Fridtjof Stoldt <fridtjof.stoldt@it.uu.se>, Tobias Wrigstad <tobias.wrigstad@it.uu.se>
Sponsor: Michael Droettboom <mdboom@gmail.com>

> Discussions-To: https://discuss.python.org/t/pep-795-revamped-deep-immutability-for-efficient-sharing-and-concurrency-safety/105356

Reviewer 4: Please also include this in a `Post-History` field right after `Python-Version`, along with the first topic:

```rst
Post-History:
  `19-Jun-2025 <https://discuss.python.org/t/pep-795-add-deep-immutability-to-python/96014`__,
  `19-Dec-2025 <https://discuss.python.org/t/pep-795-revamped-deep-immutability-for-efficient-sharing-and-concurrency-safety/105356>`__,
```

And a reminder about the reminder, please wait for the PEP to be merged before opening the main discussion topic: https://github.com/python/peps/pull/4468#issuecomment-2987903203 :)

Author: My apologies and thanks! In my defense this is the same PEP? ;) I will try to remember this from now on. 

Also, will add the Post-History now. Thanks for all the help!
